espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/espresso/EspressoW3CActionAdapter.java

```diff
@@ -119,6 +278,123 @@ public void sychronousTickActionsComplete() throws AppiumException {
+            case ZERO:
+                return KEYCODE_0;
+            case ONE:
+                return KEYCODE_1;
+            case TWO:
+                return KEYCODE_2;
+            case THREE:
+                return KEYCODE_3;
+            case FOUR:
+                return KEYCODE_4;
+            case FIVE:
+                return KEYCODE_5;
+            case SIX:
+                return KEYCODE_6;
+            case SEVEN:
+                return KEYCODE_7;
+            case EIGHT:
+                return KEYCODE_8;
+            case NINE:
+                return KEYCODE_9;
+            case ASTERISK:
+                return KEYCODE_STAR;
+            case PLUS:
+                return KEYCODE_NUMPAD_ADD;
+            case COMMA:
+                return KEYCODE_COMMA;
+            case HYPHEN:
+                return KEYCODE_MINUS;
+            case PERIOD:
+                return KEYCODE_PERIOD;
+            case FORWARD_SLASH:
+                return KEYCODE_SLASH;
+            case F1:
+                return KEYCODE_F1;
+            case F2:
+                return KEYCODE_F2;
+            case F3:
+                return KEYCODE_F3;
+            case F4:
+                return KEYCODE_F4;
+            case F5:
+                return KEYCODE_F5;
+            case F6:
+                return KEYCODE_F6;
+            case F7:
+                return KEYCODE_F7;
+            case F8:
+                return KEYCODE_F8;
+            case F9:
+                return KEYCODE_F9;
+            case F10:
+                return KEYCODE_F10;
+            case F11:
+                return KEYCODE_F11;
+            case F12:
+                return KEYCODE_F12;
+            case ZENKAKU_HANKAKU:
+                return KEYCODE_ZENKAKU_HANKAKU;
+            case SHIFT:
+                return KEYCODE_SHIFT_LEFT;
+            case CONTROL:
+                return KEYCODE_CTRL_LEFT;
+            case ALT:
+                return KEYCODE_ALT_LEFT;
+            case META:
+                return KEYCODE_META_LEFT;
+            case PAGEUP:
+                return KEYCODE_PAGE_UP;
+            case PAGEDOWN:
+                return KEYCODE_PAGE_DOWN;
+            case END:
+                return KEYCODE_MOVE_END;
+            case HOME:
+                return KEYCODE_HOME;
+            case ARROW_LEFT:
+                return KEYCODE_DPAD_LEFT;
+            case ARROW_UP:
+                return KEYCODE_DPAD_DOWN;
+            case ARROW_RIGHT:
+                return KEYCODE_DPAD_RIGHT;
+            case ARROW_DOWN:
+                return KEYCODE_DPAD_DOWN;
+            case INSERT:
+                return KEYCODE_INSERT;
+            case DELETE:
+                return KEYCODE_FORWARD_DEL;
+        }
+
+        KeyCharacterMap keyCharacterMap = KeyCharacterMap.load(KeyCharacterMap.VIRTUAL_KEYBOARD);
+        android.view.KeyEvent[] keyEvent = keyCharacterMap.getEvents(keyValue.toCharArray());
+        return keyEvent[0].getKeyCode();
```

Comment:
this is kind'a tricky. You must also consider syncing metastate, since many keys require also SHIFT to be pressed, for example uppercase stuff

Reply:
I think it should be okay. The`globalKeyInput` keeps track of the meta state (if alt, ctrl or shift is being pressed). And I'm pretty sure the 'keyValue' should be case-insensitive, because I believe we're mapping the unicode character to a 'key' and not a 'character'.

Reply:
you can try it on your own, but the actually generated key events sequence for character `H` will be:
Shift Down
Key h Down
Key h Up
Shift Up

Altogether 4 events and the first one does not contain the code of the character that we wanted

Reply:
The shift unicode that I posted maps to the shift key according to the spec. I have a switch statement that checks if it's a shift and calls the Android shift keycode.


The test I have in there confirms that shift works as expected.

Reply:
Perhaps I don't know something that you know.
Let say I execute

```java
new Actions(driver)
  .sendKeys("Hi")
  .perform()
```

this code will generate the following sequence according to https://github.com/SeleniumHQ/selenium/blob/d8f6fd34677b2662c0ac58e260efcc98f7ea3019/java/client/src/org/openqa/selenium/interactions/SendKeysAction.java

```
key down "H"
key up "H"
key down "i"
key up "i"
```

the current implementation on the server side will transform this into

```
key down SHIFT
key up SHIFT
key down "i"
key up "i"
```

Reply:
The implementation, I would expect, would do:

```
key down "H"
key up "H"
key down "i"
key up "i"
```

but it would ignore the case.

If you want uppercase H, you need to to do keydown on 'SHIFT' before keydown on 'H'.

Reply:
I assume users would expect the stuff on the client library side to simply work, shouldn't them? They don't really care about the internal implementation and nuances of the standard.